segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -552,36 +559,173 @@ private CompletableFuture<Void> writeSnapshotInfo(long snapshotId) {
+                                    }
+                                    if (!info.equals(readBackInfo)) {
+                                        throw new CompletionException(
+                                                new IllegalStateException(String.format("SystemJournal[{}] Unexpected snapshot. Expected = {} actual = {}", info, readBackInfo)));
+                                    }
+                                    return null;
+                                }, executor)
+                                .handleAsync((v, e) -> {
+                                    if (null != e) {
+                                        val ex = Exceptions.unwrap(e);
+                                        if (ex instanceof StorageNotPrimaryException) {
+                                            log.warn("SystemJournal[{}] Error while saving snapshot info to LTS {}. info = {}", containerId, snapshotInfoFileName, info, e);
+                                            throw new CompletionException(e);
+                                        }
+                                        if (attempts.incrementAndGet() > config.getMaxJournalWriteAttempts()) {
+                                            log.warn("SystemJournal[{}] Error while saving snapshot info to LTS. File = {}. info = {}", containerId, snapshotInfoFileName, info, e);
+                                            throw new CompletionException(e);
+                                        }
+                                        log.warn("SystemJournal[{}] Error while saving snapshot info to LTS. File = {}. info = {}", containerId, snapshotInfoFileName, info, e);
+                                    } else {
+                                        // No exception we are done
+                                        log.info("SystemJournal[{}] Snapshot info saved successfully. info = {}", containerId, info);
+                                        isDone.set(true);
+                                    }
+                                    return null;
+                                }, executor);
+                    }, executor);
+
+        } catch (IOException e) {
+            return CompletableFuture.failedFuture(new ChunkStorageException(snapshotInfoFileName, "Unable to serialize", e));
+        }
+    }
+
     /**
      * Find and apply latest snapshot.
      */
     private CompletableFuture<SystemSnapshotRecord> findLatestSnapshot() {
         // Step 1: Read snapshot info.
+        return readSnapshotInfo()
+            .thenComposeAsync(snapshotInfo -> {
+                if (null != snapshotInfo) {
+                    val snapshotFileName = NameUtils.getSystemJournalSnapshotFileName(containerId, snapshotInfo.getEpoch(), snapshotInfo.getSnapshotId());
+                    log.debug("SystemJournal[{}] Snapshot info read. {} pointing to {}", containerId, snapshotInfo, snapshotFileName);
+
+                    // Step 2: Validate and Read contents.
```

Comment:
Here and elsewhere in this file:
Anywhere that a section tile like "Validate and Read contents" appears, please factor this code out into a private method named `validateAndReadContents` so that the code is easy to follow here and in the method.
This replaces those 100+ line deeply nested blocks like those above with a small method several clear steps. It's a lot easier to follow.

Reply:
This seems to be the only place where we read snapshot file and parse the records. And this is already existing in its own method.

Reply:
Ok. if " Read snapshot info", "Validate and Read contents" and " Deserialize and return." don't mean anything and don't need to be communicated then delete the comments. If one or more of them do, make it a method.